### ISSUE TITLE
Error while streaming - Row conversion failed - Invalid Avro identifier

### ISSUE THREAD

**Issue by scordeiro** [confluentinc/bottledwater-pg#53]
Bottledwater aborts when it finds a column name/id with a dot character `.`

According to the error below, it seems to be a problem/limitation of Avro (?)

I didn't find any reference to this issue anywhere else, to demo it, I used  a table named `othervalues`, the issue comes from column `row.values`.

if I drop the column or change its name to something dotless, I don't get the issue and the data is successfully loaded .

```
INFO:  Row conversion failed: relation oid=16511 name=othervalues ns=public relkind=r
           0. attrelid = 16511, attname = row.values, atttypid = 25, attlen = -1, attnum = 1, attndims = 0, atttypmod = -1, attnotnull = 0, atthasdef = 0, attisdropped = 0, attcollation = 100
           1. attrelid = 16511, attname = from, atttypid = 25, attlen = -1, attnum = 2, attndims = 0, atttypmod = -1, attnotnull = 0, atthasdef = 0, attisdropped = 0, attcollation = 100
           2. attrelid = 16511, attname = to, atttypid = 25, attlen = -1, attnum = 3, attndims = 0, atttypmod = -1, attnotnull = 0, atthasdef = 0, attisdropped = 0, attcollation = 100
           3. attrelid = 16511, attname = value, atttypid = 25, attlen = -1, attnum = 4, attndims = 0, atttypmod = -1, attnotnull = 0, atthasdef = 0, attisdropped = 0, attcollation = 100
CONTEXT:  slot "bottledwater", output plugin "bottledwater", in the change callback, associated LSN 0/********
/usr/local/bin/bottledwater: Replication stream was unexpectedly terminated: ERROR:  output_avro_change: row conversion failed: Invalid Avro identifier
CONTEXT:  slot "bottledwater", output plugin "bottledwater", in the change callback, associated LSN 0/********
```

This introduces a strong limitation since PostgreSQL allows [for any quoted identifiers to contain any character, (except the character with code zero)](http://www.postgresql.org/docs/current/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS).

**Comment by ept** [comment]
Thanks for the report! I guess we'll need to replace the dots to make Avro happy. A simple solution may be to replace anything that's not alphanumeric or underscore with an underscore. (Avro names must match the regex `[a-zA-Z_][a-zA-Z0-9_]*`.) Do you want to try making a pull request?

**Comment by scordeiro** [comment]
@ept Thanks for the quick reply. We're talking about dot characters now but someone may mention other characters in the future which may not be in Avro's limitation.

If you replace any non `[a-zA-Z_][a-zA-Z0-9_]` by an underscore, it could interfere with a table having a column with similar names (although unlikely to happen much).
As an example, let's say that, for whatever reason, a table has columns "value@row", "value.row" and "value?row", if you replace each unusual character with an underscore you'll probably end up with a conflict.

Also, you'd end up with non-matching column names in Kafka against the database's.

I read a bit about Avro's names, and I'm now able to better understand this limitation, I guess I'll have to restrain from using special characters in the column names for the moment.

As for the PR, if I find the time I'll do it, right now I'm only using bottledwater for some tests, it's not critical and its code is complex enough to require a fair amount of time to analyse it.

**Comment by ept** [comment]
Yes, name collisions would need to be handled if we do some character substitution. That applies not just to column names, but also table names (and potentially tables with the same name in different Postgres schemas). We will have to solve this sooner or later, but as you say, the easiest workaround for now is to just use alphanumeric column names.

**Comment by samstokes** [comment]
Proposed fix (in #98):

Sanitise identifiers to be acceptible to Avro using an encoding similar to the "percent encoding" used in URLs.  [Unsupported characters](https://avro.apache.org/docs/1.8.1/spec.html#names) are replaced by a hexadecimal representation: e.g. "person.name" -> "person_2e_name"

A couple of gotchas:
 * This encoding is not entirely unambiguous, since:
          "person_2e_name" -> "person_2e_name".
   This could be worked around by treating '_' as an invalid character and encoding it too (e.g. "person_2e_name" -> "person_5f_2e_5f_name"), but that seems ugly, especially since we ourselves generate names like "<relname>_pkey".
 * This encoding is not Unicode-aware!  Given a string containing (bytes representing the encoded form of) non-ASCII characters, it should still return a valid Avro identifier, but its behaviour is otherwise unspecified.

**Comment by samstokes** [comment]
Thanks to @uhoh-itsmaciek I discovered and fixed a [bug in my sanitisation function](https://github.com/samstokes/bottledwater-pg/commit/39ecbc6397143cbcec739043b274813be45cdec1), so now the Unicode caveat above no longer applies.

The encoding of non-ASCII characters isn't particularly intuitive, but it is deterministic: it's the underscore encoding of the bytes representing those characters in the server encoding (which I'm guessing we never change from UTF-8).  e.g.: "crêpes" -> "cr_c3__aa_pes"
